storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3108,10 +3108,1655 @@ buf_block_t *fsp_get_latched_xdes_page(
+    if (!flst_node_offset_validate(&next_addr))
+      return DB_CORRUPTION;
+  }
+
+  if (prev_addr.page >= free_limit)
+  {
+    if (prev_addr.page != FIL_NULL)
+      return DB_CORRUPTION;
+  }
+
+  if (prev_addr.page != FIL_NULL && !flst_node_offset_validate(&prev_addr))
+    return DB_CORRUPTION;
+
+  if (prev_addr.page != FIL_NULL)
+  {
+    *prev_block= fsp_get_latched_page(page_id_t{0, prev_addr.page}, mtr, &err);
+    if (err) return err;
+  }
+
+  return DB_SUCCESS;
+}
+
+/** Complete the steps for removing the file list node
+@param base       base block where free list starts
+@param boffset    offset where list starts
+@param descr      descriptor to be removed
+@param mtr        mini-transaction */
+void flst_remove_complete(buf_block_t *base, uint16_t boffset,
+                          xdes_t *descr, mtr_t *mtr) noexcept
+{
+  fil_addr_t next_addr= flst_get_next_addr(descr + XDES_FLST_NODE);
+
+  fil_addr_t prev_addr= flst_get_prev_addr(descr + XDES_FLST_NODE);
+
+  byte *list= base->page.frame + boffset;
+
+  buf_block_t *prev_block= nullptr;
+  buf_block_t *next_block= nullptr;
+
+  if (prev_addr.page != FIL_NULL)
+  {
+    prev_block= mtr->get_already_latched(page_id_t{0, prev_addr.page},
+                                         MTR_MEMO_PAGE_SX_FIX);
+    ut_ad(prev_block);
+  }
+
+  if (next_addr.page != FIL_NULL)
+  {
+    next_block= mtr->get_already_latched(page_id_t{0, next_addr.page},
+                                         MTR_MEMO_PAGE_SX_FIX);
+    ut_ad(next_block);
+  }
+
+  if (!prev_block)
+    flst_write_addr(*base, list + FLST_FIRST,
+                    next_addr.page, next_addr.boffset, mtr);
+  else
+    flst_write_addr(*prev_block,
+                    prev_block->page.frame
+                      + prev_addr.boffset + FLST_NEXT,
+                    next_addr.page, next_addr.boffset, mtr);
+
+  if (next_addr.page == FIL_NULL)
+    flst_write_addr(*base, list + FLST_LAST,
+                    prev_addr.page, prev_addr.boffset, mtr);
+  else
+    flst_write_addr(*next_block,
+                    next_block->page.frame
+		      + next_addr.boffset + FLST_PREV,
+                    prev_addr.page, prev_addr.boffset, mtr);
+
+  /* Decrement the FSP_FREE length in FSP_HEADER */
+  byte *len= list + FLST_LEN;
+  mtr->write<4>(*base, len, mach_read_from_4(len) - 1);
+}
+
+/** Prepare the steps for adding the block into last of the list
+@param base             block where list starts
+@param boffset          offset to find the list
+@param free_limit       maximum free limit in the tablespace
+@param last_block_list  last block in the list
+@param mtr              mini-transaction
+@return error code */
+dberr_t flst_add_last_prepare(buf_block_t *base, uint16_t boffset,
+                              uint32_t free_limit,
+                              buf_block_t **last_block_list,
+                              mtr_t *mtr) noexcept
+{
+  if (!flst_get_len(base->page.frame + boffset))
+    return DB_SUCCESS;
```

Comment:
It is unclear how the case of an empty list would be handled in `flst_add_last_complete()` or its callers. A comment about that would be helpful.

Would `ut_ad(!*last_block_list)` be satistified at the start of this function? If yes, I would suggest adding that assertion to document that the output parameter has been initialized by the caller.